### PR TITLE
Homeapis

### PR DESCRIPTION
**Pull request by Chandini-212634**

### CODE REVIEW

File: src/main/java/com/ust/sourcecourse/configuration/service/HomeService.java [hunk 1]
```diff
@@ -7,52 +7,83 @@
...
+
+	private DBTableColumn getDBTableColumn(SourceColumn sourceColumn) {
+		DBTableColumnMetadata metadata = DBTableColumnMetadata.builder().type(sourceColumn.getType())
+				.isPrimary(sourceColumn.isPrimary()).isNullable(sourceColumn.isNullable())
+				.isUnique(sourceColumn.isUnique()).defaultValue(sourceColumn.getDefaultValue()).build();
+		return DBTableColumn.builder().uid(sourceColumn.getUid()).name(sourceColumn.getName())
+				.description(sourceColumn.getDescription()).metadata(metadata).tags(sourceColumn.getTags()).build();
+	}
+
+	public DBTableColumn UpdateDescriptionAndRemoveAllTags(Long uid, HomeRequest request) {
```
Comment (shijukbabu-ust) [inline src/main/java/com/ust/sourcecourse/configuration/service/HomeService.java:67]: Where is this used? If not used, remove it.

File: src/main/java/com/ust/sourcecourse/configuration/service/HomeService.java [hunk 2]
```diff
@@ -7,52 +7,83 @@
...
+		sourceColumn.setDescription(request.getDescription());
+		sourceColumn.getTags().clear(); // Remove all tags
+		SourceColumn updatedSourceColumn = sourceColumnRepository.save(sourceColumn);
+		return getDBTableColumn(updatedSourceColumn);
+	}
+
+	public void deleteTableTags(Long uid) {
+		SourceTable sourceTable = sourceTableRepository.findById(uid)
+				.orElseThrow(() -> new EntityNotFoundException("Table not found"));
+		sourceTable.getTags().clear();
```
Comment (shijukbabu-ust) [inline src/main/java/com/ust/sourcecourse/configuration/service/HomeService.java:78]: Why deleting all tags. API should be to remove list of tags. Not remove all tags. 
Request payload should have list of tags to be removed

File: src/main/java/com/ust/sourcecourse/configuration/service/HomeService.java [hunk 3]
```diff
@@ -7,52 +7,83 @@
...
-//			existingProject.setDescription(project.getDescription());
-//			existingProject.setModifiedBy(project.getModifiedBy());
-//			existingProject.setModifiedTimestamp(LocalDateTime.now());
-//			return projectRepository.save(existingProject);
-//		}
-//		
-//		public void deleteProject(Long id) {
-//			projectRepository.deleteById(id);
-//		}
+	public void deleteColumnTags(Long uid) {
```
Comment (shijukbabu-ust) [inline src/main/java/com/ust/sourcecourse/configuration/service/HomeService.java:82]: Similar to above, 
Why deleting all tags. API should be to remove list of tags. Not remove all tags. 
Request payload should have list of tags to be removed

**Review by shijukbabu-ust — CHANGES_REQUESTED** [review]

File: src/main/java/com/ust/sourcecourse/configuration/service/DBDataSourceService.java [hunk 4]
```diff
@@ -69,17 +78,86 @@ private DBTable getDBTable(SourceTable sourceTable) {
...
+			return table.get().getTags();
+		} else {
+			throw new EntityNotFoundException("SourceTable with uid " + uid + " not found");
+		}
+	}
+
+	public SourceColumn updateColumnDescriptionAndTags(Long uid, HomeRequest request) {
+		Optional<SourceColumn> column = sourceColumnRepository.findById(uid);
+		if (column.isPresent()) {
+			column.get().setTags(request.getTags());
```
Comment (shijukbabu-ust) [inline src/main/java/com/ust/sourcecourse/configuration/service/DBDataSourceService.java:122]: You need to update description as well. 
Also, this is an update. So existing tags should not be removed. Setting tags directly from request will remove existing tags. Also, we need to handle duplicates. 

You can refer this https://github.com/sourcecoursepoc/source-course-configuration/blob/dev/src/main/java/com/ust/sourcecourse/configuration/service/ProjectGroupService.java#L128

File: src/main/java/com/ust/sourcecourse/configuration/service/DBDataSourceService.java [hunk 5]
```diff
@@ -69,17 +78,86 @@ private DBTable getDBTable(SourceTable sourceTable) {
...
+			sourceColumnRepository.save(column.get());
+			return column.get();
+		} else {
+			throw new EntityNotFoundException("SourceColumn with uid " + uid + " not found");
+		}
+	}
+
+	public SourceTable updateTableDescriptionAndTags(Long uid, HomeRequest request) {
+		Optional<SourceTable> table = sourceTableRepository.findById(uid);
```
Comment (shijukbabu-ust) [inline src/main/java/com/ust/sourcecourse/configuration/service/DBDataSourceService.java:131]: Same as above. Refer https://github.com/sourcecoursepoc/source-course-configuration/blob/dev/src/main/java/com/ust/sourcecourse/configuration/service/ProjectGroupService.java#L128

File: src/main/java/com/ust/sourcecourse/configuration/service/DBDataSourceService.java [hunk 6]
```diff
@@ -69,17 +78,86 @@ private DBTable getDBTable(SourceTable sourceTable) {
...
+		} else {
+			throw new EntityNotFoundException("SourceTable with uid " + uid + " not found");
+		}
+	}
+
+	public void deleteColumnTags(Long uid) {
+		Optional<SourceColumn> column = sourceColumnRepository.findById(uid);
+		if (column.isPresent()) {
+			List<String> Tags=column.get().getTags();
+			sourceColumnRepository.delete((SourceColumn) Tags);
```
Comment (shijukbabu-ust) [inline src/main/java/com/ust/sourcecourse/configuration/service/DBDataSourceService.java:145]: This is removing the entity, not tags. Also you need to get the tag as input which needs to be removed. Refer
https://github.com/sourcecoursepoc/source-course-configuration/blob/dev/src/main/java/com/ust/sourcecourse/configuration/service/ProjectGroupService.java#L150

**Review by shijukbabu-ust — CHANGES_REQUESTED** [review]

File: src/main/java/com/ust/sourcecourse/configuration/controller/DBDataSourceController.java [hunk 7]
```diff
@@ -42,4 +46,40 @@ public ResponseEntity<List<DBDataSourceInfo>> getDBSourceInfo() {
...
+		return ResponseEntity.ok(columnTags);
+	}
+
+	@GetMapping("/tags/tables/{uid}")
+	public ResponseEntity<List<String>> getTableTags(@PathVariable Long uid) {
+		List<String> tableTags = dataSourceService.getTableTags(uid);
+		return ResponseEntity.ok(tableTags);
+	}
+
+	@PutMapping("/tags/columns/{uid}")
```
Comment (shijukbabu-ust) [inline src/main/java/com/ust/sourcecourse/configuration/controller/DBDataSourceController.java:61]: Since we are updating tags and description, url need not have "tags" path. Pls remove

File: src/main/java/com/ust/sourcecourse/configuration/controller/DBDataSourceController.java [hunk 8]
```diff
@@ -42,4 +46,40 @@ public ResponseEntity<List<DBDataSourceInfo>> getDBSourceInfo() {
...
+	public ResponseEntity<List<String>> updateColumnTags(@PathVariable Long uid, @RequestBody HomeRequest request) {
+		SourceColumn column = dataSourceService.updateColumnDescriptionAndTags(uid, request);
+		return ResponseEntity.ok(column.getTags());
+	}
+
+	@PutMapping("/tags/tables/{uid}")
```
Comment (shijukbabu-ust) [inline src/main/java/com/ust/sourcecourse/configuration/controller/DBDataSourceController.java:67]: Similar to above, Since we are updating tags and description, url need not have "tags" path. Pls remove

**Review by shijukbabu-ust — CHANGES_REQUESTED** [review]

File: src/main/java/com/ust/sourcecourse/configuration/controller/HomeController.java [hunk 9]
```diff
@@ -23,4 +29,43 @@ public ResponseEntity<List<HomePageResponse>> getHomePageDetails() {
 		return ResponseEntity.ok(homeService.getHomePageDetails());
 	}
 
+
+	@GetMapping("/tags/columns{uid}")
+	public ResponseEntity<List<String>> getDescriptionAndTags1(@PathVariable Long uid) {
+		List<String> homePageResponse = homeService.getDescriptionAndTags(uid);
```
[the remark this answers is not in the document]
Comment (Chandini-212634) [reply src/main/java/com/ust/sourcecourse/configuration/controller/HomeController.java]: resolved

File: src/main/java/com/ust/sourcecourse/configuration/service/HomeService.java [hunk 10]
```diff
@@ -7,52 +7,83 @@
...
+			HomePageResponse response = HomePageResponse.builder().projectName(project.getName())
+					.description(project.getDescription()).build();
+			response.setTables(project.getProjectTables().size());
+			response.setGroups(project.getGroups().size());
+			homePageResponses.add(response);
 		}
+		return homePageResponses;
+	}
+
+	public List<String> getDescriptionAndTags(Long uid) {
```
[the remark this answers is not in the document]
Comment (Chandini-212634) [reply src/main/java/com/ust/sourcecourse/configuration/service/HomeService.java]: resolved

[the remark this answers is not in the document]
Comment (Chandini-212634) [reply src/main/java/com/ust/sourcecourse/configuration/service/HomeService.java]: resolved

File: src/main/java/com/ust/sourcecourse/configuration/service/HomeService.java [hunk 11]
```diff
@@ -7,52 +7,83 @@
...
+		SourceTable homePage = sourceTableRepository.findById(uid).orElseThrow();
+
+		return homePage.getTags();
+	}
+
+	public DBTableColumn UpdateDescriptionandTags(Long uid, HomeRequest request) {
```
[the remark this answers is not in the document]
Comment (Chandini-212634) [reply src/main/java/com/ust/sourcecourse/configuration/service/HomeService.java]: resolved

File: src/main/java/com/ust/sourcecourse/configuration/service/HomeService.java (hunk 3 above)
Comment (Chandini-212634) [reply src/main/java/com/ust/sourcecourse/configuration/service/HomeService.java]: resolved

File: src/main/java/com/ust/sourcecourse/configuration/controller/DBDataSourceController.java [hunk 12]
```diff
@@ -42,4 +46,40 @@ public ResponseEntity<List<DBDataSourceInfo>> getDBSourceInfo() {
...
+	public ResponseEntity<List<String>> updateTableTags(@PathVariable Long uid, @RequestBody HomeRequest request) {
+		SourceTable table = dataSourceService.updateTableDescriptionAndTags(uid, request);
+		return ResponseEntity.ok(table.getTags());
+	}
+
+	@DeleteMapping("/tags/columns/{uid}")
```
[the remark this answers is not in the document]
Comment (Chandini-212634) [reply src/main/java/com/ust/sourcecourse/configuration/controller/DBDataSourceController.java]: resolved

File: src/main/java/com/ust/sourcecourse/configuration/request/HomeRequest.java [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+
+import java.util.List;
+
+import jakarta.validation.constraints.NotBlank;
+import lombok.Data;
+@Data
+public class HomeRequest {
+	@NotBlank
+	private String description;
+	@NotBlank
```
[the remark this answers is not in the document]
Comment (Chandini-212634) [reply src/main/java/com/ust/sourcecourse/configuration/request/HomeRequest.java]: resloved

File: src/main/java/com/ust/sourcecourse/configuration/service/DBDataSourceService.java (hunk 4 above)
Comment (Chandini-212634) [reply src/main/java/com/ust/sourcecourse/configuration/service/DBDataSourceService.java]: resolved

File: src/main/java/com/ust/sourcecourse/configuration/controller/DBDataSourceController.java (hunk 7 above)
Comment (Chandini-212634) [reply src/main/java/com/ust/sourcecourse/configuration/controller/DBDataSourceController.java]: resolved

File: src/main/java/com/ust/sourcecourse/configuration/controller/DBDataSourceController.java [hunk 14]
```diff
@@ -42,4 +44,59 @@ public ResponseEntity<List<DBDataSourceInfo>> getDBSourceInfo() {
...
+		return ResponseEntity.ok(sourcetables);
+	}
+
+	@GetMapping("/column/search/{tag}")
+	public ResponseEntity<List<DBTableColumn>> searchColumnByTag(@PathVariable String tag) {
+		List<DBTableColumn> groups = dataSourceService.searchcolumnByTag(tag);
+		return ResponseEntity.ok(groups);
+	}
+	@PostMapping("table/post/{tags}")
+    public ResponseEntity<List<String>> addTagToSourceTable(@PathVariable ("id") Long uid, @RequestBody  List<String> tags) {
```
Comment (shijukbabu-ust) [inline src/main/java/com/ust/sourcecourse/configuration/controller/DBDataSourceController.java:91]: This should be accepting TagDescriptionRequest, not just tags alone. Updating description also should be possible

File: src/main/java/com/ust/sourcecourse/configuration/controller/DBDataSourceController.java (hunk 8 above)
Comment (Chandini-212634) [reply src/main/java/com/ust/sourcecourse/configuration/controller/DBDataSourceController.java]: resloved

File: src/main/java/com/ust/sourcecourse/configuration/controller/DBDataSourceController.java [hunk 15]
```diff
@@ -42,4 +44,59 @@ public ResponseEntity<List<DBDataSourceInfo>> getDBSourceInfo() {
...
+        List<String> updatedSourceTable= dataSourceService.addTagSourceTable(uid, tags);
+        return ResponseEntity.ok(updatedSourceTable);
+    }
+	@PostMapping("column/post/{tags}")
+    public ResponseEntity<List<String>> addTagToSourceColumn(@PathVariable ("id") Long uid, @RequestBody  List<String> tags) {
```
Comment (shijukbabu-ust) [inline src/main/java/com/ust/sourcecourse/configuration/controller/DBDataSourceController.java:96]: This should be accepting TagDescriptionRequest, not just tags alone. Updating description also should be possible

File: src/main/java/com/ust/sourcecourse/configuration/controller/DBDataSourceController.java (hunk 14 above)
Comment (shijukbabu-ust) [reply src/main/java/com/ust/sourcecourse/configuration/controller/DBDataSourceController.java:91]: No need of post in URL

File: src/main/java/com/ust/sourcecourse/configuration/controller/DBDataSourceController.java (hunk 15 above)
Comment (shijukbabu-ust) [reply src/main/java/com/ust/sourcecourse/configuration/controller/DBDataSourceController.java:96]: No need of POST in URL

File: src/main/java/com/ust/sourcecourse/configuration/exceptions/ResourceNotFoundException.java [hunk 16]
```diff
@@ -2,6 +2,23 @@
 
 import java.util.List;
 
+<<<<<<< HEAD
```
Comment (shijukbabu-ust) [inline src/main/java/com/ust/sourcecourse/configuration/exceptions/ResourceNotFoundException.java:5]: Resolve the conflicts and commit

File: src/main/java/com/ust/sourcecourse/configuration/repository/SourceColumnRepository.java [hunk 17]
```diff
@@ -1,10 +1,16 @@
...
+
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.stereotype.Repository;
 
 import com.ust.sourcecourse.configuration.entity.SourceColumn;
 @Repository
 public interface SourceColumnRepository extends JpaRepository<SourceColumn, Long> {
 
-}
+	List<String> getTags();
```
Comment (shijukbabu-ust) [inline src/main/java/com/ust/sourcecourse/configuration/repository/SourceColumnRepository.java:12]: Why is this required?

File: src/main/java/com/ust/sourcecourse/configuration/repository/SourceTableRepository.java [hunk 18]
```diff
@@ -1,10 +1,16 @@
...
 
+import java.util.List;
+
 import org.springframework.data.jpa.repository.JpaRepository;
 
 import com.ust.sourcecourse.configuration.entity.SourceTable;
 
 public interface SourceTableRepository extends JpaRepository<SourceTable, Long> {
 
+	List<String> getTags();
```
Comment (shijukbabu-ust) [inline src/main/java/com/ust/sourcecourse/configuration/repository/SourceTableRepository.java:11]: Why is this required?

File: src/main/java/com/ust/sourcecourse/configuration/request/TagDescriptionRequest.java [hunk 19]
```diff
@@ -0,0 +1,15 @@
+package com.ust.sourcecourse.configuration.request;
+
+import java.util.List;
+
+import jakarta.validation.constraints.NotBlank;
+import lombok.Data;
+@Data
+public class TagDescriptionRequest {
+	@NotBlank
```
Comment (shijukbabu-ust) [inline src/main/java/com/ust/sourcecourse/configuration/request/TagDescriptionRequest.java:9]: @NotBlank is not required.

**Review by shijukbabu-ust — CHANGES_REQUESTED** [review]

File: src/main/java/com/ust/sourcecourse/configuration/controller/DBDataSourceController.java (hunk 14 above)
Comment (Chandini-212634) [reply src/main/java/com/ust/sourcecourse/configuration/controller/DBDataSourceController.java]: resolved

File: src/main/java/com/ust/sourcecourse/configuration/controller/DBDataSourceController.java (hunk 15 above)
Comment (Chandini-212634) [reply src/main/java/com/ust/sourcecourse/configuration/controller/DBDataSourceController.java]: resolved

File: src/main/java/com/ust/sourcecourse/configuration/exceptions/ResourceNotFoundException.java (hunk 16 above)
Comment (Chandini-212634) [reply src/main/java/com/ust/sourcecourse/configuration/exceptions/ResourceNotFoundException.java]: resolved

File: src/main/java/com/ust/sourcecourse/configuration/repository/SourceColumnRepository.java (hunk 17 above)
Comment (Chandini-212634) [reply src/main/java/com/ust/sourcecourse/configuration/repository/SourceColumnRepository.java]: resolved

File: src/main/java/com/ust/sourcecourse/configuration/repository/SourceTableRepository.java (hunk 18 above)
Comment (Chandini-212634) [reply src/main/java/com/ust/sourcecourse/configuration/repository/SourceTableRepository.java]: resolved

File: src/main/java/com/ust/sourcecourse/configuration/request/TagDescriptionRequest.java [hunk 20]
```diff
@@ -0,0 +1,14 @@
...
+import java.util.List;
+
+import lombok.Data;
+
+@Data
+public class TagDescriptionRequest {
+
+	private String description;
+
+	private List<String> Tags;
```
Comment (shijukbabu-ust) [inline src/main/java/com/ust/sourcecourse/configuration/request/TagDescriptionRequest.java:12]: Follow Java naming conventions. Variable name should start with lower case letters. 
Change to tags and change everywhere accordingly.

**Review by shijukbabu-ust — CHANGES_REQUESTED** [review]

File: src/main/java/com/ust/sourcecourse/configuration/request/TagDescriptionRequest.java (hunk 20 above)
Comment (Chandini-212634) [reply src/main/java/com/ust/sourcecourse/configuration/request/TagDescriptionRequest.java]: resolved

File: src/main/resources/application.properties [hunk 21]
```diff
@@ -1,9 +1,17 @@
-server.port=8080
-server.servlet.context-path=/sourcecourse
-spring.datasource.url=jdbc:mysql://localhost:3306/sourcecourse?useSSL =true
+server.port=8081
```
Comment (shijukbabu-ust) [inline src/main/resources/application.properties:1]: Do not change this file.

Comment (Chandini-212634) [reply src/main/resources/application.properties]: resolved

**Review by shijukbabu-ust — APPROVED** [review]
